src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs

```diff
@@ -12,13 +16,72 @@ internal static class WebInputSources
         /// </summary>
         public static ImmutableHashSet<SourceInfo> SourceInfos { get; }
 
+        /// <summary>
+        /// Cached information if the specified symbol is a Asp.Net Core Controller: (compilation) -> ((class symbol) -> (is Controller))
+        /// </summary>
+        private static readonly BoundedCacheWithFactory<Compilation, ConcurrentDictionary<ISymbol, bool>> s_classIsControllerByCompilation =
+            new BoundedCacheWithFactory<Compilation, ConcurrentDictionary<ISymbol, bool>>();
+
         /// <summary>
         /// Statically constructs.
         /// </summary>
         static WebInputSources()
         {
             var sourceInfosBuilder = PooledHashSet<SourceInfo>.GetInstance();
 
+            sourceInfosBuilder.AddSourceInfo(
+                // checking all System.Object derived types is expensive, so it first checks if MicrosoftAspNetCoreMvcControllerBase is resolvable
+                WellKnownTypeNames.MicrosoftAspNetCoreMvcControllerBase,
+                WellKnownTypeNames.SystemObject,
+                 new ParameterMatcher[]{
+                    (parameter, wellKnownTypeProvider) => {
+                        if (!(parameter.ContainingSymbol is IMethodSymbol methodSymbol)
+                            || !(methodSymbol.ContainingSymbol is INamedTypeSymbol typeSymbol))
+                        {
+                            return false;
+                        }
+
+                        var classCache = s_classIsControllerByCompilation.GetOrCreateValue(wellKnownTypeProvider.Compilation, (compilation) => new ConcurrentDictionary<ISymbol, bool>());
+                        if (!classCache.TryGetValue(methodSymbol.ContainingSymbol, out bool isController))
+                        {
+                            if ((!typeSymbol.GetBaseTypesAndThis().Any(x => x.Name.EndsWith("Controller", System.StringComparison.Ordinal))
```

Comment:
Nit: Could only add to `s_classIsControllerByCompilation` when the entry isn't found:

```csharp
                            if ((!typeSymbol.GetBaseTypesAndThis().Any(x => x.Name.EndsWith("Controller", System.StringComparison.Ordinal))
                                && (!wellKnownTypeProvider.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftAspNetCoreMvcControllerAttribute, out var controllerAttributeTypeSymbol)
                                    || !typeSymbol.HasDerivedTypeAttribute(controllerAttributeTypeSymbol)))
                                || !wellKnownTypeProvider.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftAspNetCoreMvcNonControllerAttribute, out var nonControllerAttributeTypeSymbol)
                                || typeSymbol.HasDerivedTypeAttribute(nonControllerAttributeTypeSymbol))
                            {
                                isController = false;
                            }
                            else
                            {
                                isController = true;
                            }

                            classCache.TryAdd(typeSymbol, isController);
```